mala/datahandling/data_shuffler.py

```diff
@@ -0,0 +1,138 @@
+                False
+
+    def add_snapshot(self, input_npy_file, input_npy_directory,
+                     output_npy_file, output_npy_directory,
+                     add_snapshot_as="te",
+                     output_units="1/(eV*A^3)", input_units="None",
+                     calculation_output_file="", snapshot_type="numpy"):
+        super(DataShuffler, self).\
+            add_snapshot(input_npy_file, input_npy_directory,
+                         output_npy_file, output_npy_directory,
+                         add_snapshot_as="te",
+                         output_units=output_units, input_units=input_units,
+                         calculation_output_file=calculation_output_file,
+                         snapshot_type=snapshot_type)
+
+    def shuffle_snapshots(self, save_path, save_name="mala_shuffled_snapshot*",
+                          shuffle_dimensions=None):
+        # Check the dimensions of the snapshots.
+        self._check_snapshots()
+        snapshot_size_list = [snapshot.grid_size for snapshot in
+                              self.parameters.snapshot_directories_list]
+
+        # Calculate how data will be split.
+        if shuffle_dimensions is None:
+            # If all snapshots have the same size, we will just re-use
+            # this size if not otherwise specified.
+            if np.max(snapshot_size_list) == np.min(snapshot_size_list):
+                shuffle_dimensions = self.parameters.\
+                    snapshot_directories_list[0].grid_dimension
+            else:
+                # If they do have different sizes, we start with the smallest
+                # snapshot, there is some padding down below anyhow.
+                shuffle_dimensions = self.parameters.\
+                    snapshot_directories_list[np.argmin(snapshot_size_list)].\
+                    grid_dimension
+
+        # Padding the new snapshot size until we can fit all snapshots.
+        number_of_data_points = np.sum(snapshot_size_list)
+        if number_of_data_points % \
+                int(np.prod(shuffle_dimensions)) != 0:
+            old_shuffle_dimensions = shuffle_dimensions.copy()
+            while number_of_data_points % \
+                    int(np.prod(shuffle_dimensions)) != 0:
+                shuffle_dimensions[3] += 1
+            printout("Had to readjust shuffle dimensions from",
+                     old_shuffle_dimensions, "to", shuffle_dimensions)
+
+        number_of_new_snapshots = int(number_of_data_points /
+                                      int(np.prod(shuffle_dimensions)))
+        printout("Data shuffler will generate", number_of_new_snapshots,
+                 "new snapshots.")
+
+        # Prepare permutations.
+        permutations = []
+        for i in range(0, number_of_new_snapshots):
+            permutations.append(np.random.permutation(
+                int(np.prod(shuffle_dimensions))))
+
+        # Load the data (via memmap).
```

Comment:
Loading the data all at once and using mmap to accommodate large data sets is nice and flexible but I worry the chunking and randomizing process below will be very slow if it's accessing data on disk since there's no guarantee our slicing/chunking will align with the way the data is stored on the disk.

**Options**
1. Leave it as is and test how it handles a large (maybe around 1TB) data set -- maybe you've already done this?
2. Separate intra and inter snapshot shuffling and use intermediate writing/reading to/from disk to reduce memory footprint to the size of the single largest snapshot provided. This will double the amount of read/write from disk but may still be way faster because we will prevent manipulating arrays in disk (as long as the provided snapshots are individually smaller than the available memory).
---> Calculate dimensions of chunks we need to make of each input snapshot.
---> **Loop thru input snapshots**:   Shuffle and chunk each input snapshot, then write to disk as chunks.
---> **Loop thru output snapshots**: Read chunks from each input snapshot, reconstitute into output snapshot, shuffle again, write to disk as output snapshot
---> Delete intermediate files on disk

Reply:
So I have tested this on the Al data and it worked quite well I got to say. For 13 snapshots (each with the 200x200x200 grid, so hundreds of GBs) the shuffling took only a few minutes. I think this is an ok timing, since this step will only be done once, right?

Reply:
That sounds like great timing. Im curious what kind of hardware this was with? Was it able to parallelize?

Reply:
To be precise: It was 20 minutes for 12 Al snapshots. 
What timing did you observe? I don't have hardware specifics right now, but I know from prior experience that the hard drives on our cluster are on the faster side. 
This was not parallelized, but by using OpenPMD we should be able to parallelize this step with relative ease in the future.

Reply:
That seems pretty good. The shuffling alone took more than 30 minutes for 8 SiO2 snapshots (2.6e7 examples) -- including the loading and writing would at least double that.